### PR TITLE
Add api call to get payment methods

### LINKED ISSUE #460: payments: api call for getting payment providers with currency conversion ratios
Per payment provider we need to be able to define multiple currencies. Per provider, an object should be returned to the client. These will be used to show a list of options that can be used to pay when paying via the pay widget

properties:

- logo_url
- provider_id
- embedded_app
- methods:
  - currency
  - amount
  - precision

On the pay widget:
- add 'base payment method'
  - currency,  provider
use this method to calculate the amount of the other methods if the boolean 'calculate_amount' on the form settings is true (true by default)

On payment provider model, store the conversion ratio:

```
conversion_ratio =  {
   'base': 'EUR',
   'values': [{'currency': 'USD', 'amount': 1.21}, {'currency': 'TFT', 'amount': 10.0}]
}
```

### PR DESCRIPTION
**Pull request by lucasvanhalst**
Closes #460

### CODE REVIEW

File: src/rogerthat/to/payment.py [hunk 1]
```diff
@@ -595,3 +597,27 @@ def __init__(self, success=False, error=None, result=None):
...
+    target = unicode_property('target')
+    test_mode = bool_property('test_mode')
+    base_method = typed_property('base_method', BasePaymentMethod)  # type: BasePaymentMethod
+    methods = typed_property('methods', PaymentMethodTO, True)  # type: list[PaymentMethodTO]
+
+
+class PaymentProviderMethodsTO(TO):
+    logo_url = unicode_property('logo_url')
+    provider_id = unicode_property('provider_id')
+    embedded_app = unicode_property('embedded_app')
```
Comment (RubenMattan) [inline src/rogerthat/to/payment.py]: Lets rename to embedded_app_id

File: src/rogerthat/to/payment.py [hunk 2]
```diff
@@ -595,3 +597,27 @@ def __init__(self, success=False, error=None, result=None):
...
+class GetPaymentMethodsRequestTO(TO):
+    target = unicode_property('target')
+    test_mode = bool_property('test_mode')
+    base_method = typed_property('base_method', BasePaymentMethod)  # type: BasePaymentMethod
+    methods = typed_property('methods', PaymentMethodTO, True)  # type: list[PaymentMethodTO]
+
+
+class PaymentProviderMethodsTO(TO):
+    logo_url = unicode_property('logo_url')
+    provider_id = unicode_property('provider_id')
```
Comment (RubenMattan) [inline src/rogerthat/to/payment.py]: Lets refactor to provider:{'id': xxx, 'logo': xxx, 'name': xxx}

**Comment by RubenMattan** [comment]
Also got multiple unit tests failing now:
```
Traceback (most recent call last):
  File "/Users/ruben/mobicage/git/rogerthat-backend/src-test/rogerthat_tests/mobicage/bizz/test_messaging.py", line 647, in testGeneratePushJSON2
    create_service_identity(svc_user, to)
  File "/Users/ruben/mobicage/git/rogerthat-backend/src/mcfw/rpc.py", line 164, in typechecked_return
    result = f(*args, **kwargs)
  File "/Users/ruben/mobicage/git/rogerthat-backend/src/mcfw/rpc.py", line 142, in typechecked_f
    return f(**kwargs)
  File "/Users/ruben/mobicage/git/rogerthat-backend/src/rogerthat/bizz/service/__init__.py", line 844, in create_service_identity
    app_id = to.app_ids[0] if to.app_ids is not MISSING and to.app_ids else get_default_app().app_id
AttributeError: 'NoneType' object has no attribute 'app_id'
```

```
Traceback (most recent call last):
  File "/Users/ruben/mobicage/git/rogerthat-backend/src-test/rogerthat_tests/mobicage/bizz/test_mfd.py", line 522, in testFormMessages
    self.assertStatusValid(mfd)
  File "/Users/ruben/mobicage/git/rogerthat-backend/src-test/rogerthat_tests/mobicage/bizz/test_mfd.py", line 405, in assertStatusValid
    self.assertEqual(mfd.status, MessageFlowDesign.STATUS_VALID, msg)
AssertionError: MFD: A
    status: BROKEN
    error: {&apos;reason&apos;: &apos;Base method currency is a required field&apos;, &apos;property&apos;: &apos;base_method&apos;} - Invalid value For Pay '15'
    broken subflows: []
```

```
Traceback (most recent call last):
  File "/Users/ruben/mobicage/git/rogerthat-backend/src-test/rogerthat_tests/mobicage/bizz/test_mfd.py", line 584, in testSaveAllModulesinFlow
    self.assert_(mfd.xml)
AssertionError: None is not true
```

```
Traceback (most recent call last):
  File "/Users/ruben/.p2/pool/plugins/org.python.pydev_6.2.0.201711281614/pysrc/_pydev_runfiles/pydev_runfiles.py", line 468, in __get_module_from_str
    mod = __import__(modname)
  File "/Users/ruben/mobicage/git/rogerthat-backend/src-test/rogerthat_tests/test_monkey_patch.py", line 21, in <module>
    from rogerthat.rpc import users
  File "/Users/ruben/mobicage/git/rogerthat-backend/src/rogerthat/rpc/users.py", line 25, in <module>
    from rogerthat.utils import slog
  File "/Users/ruben/mobicage/git/rogerthat-backend/src/rogerthat/utils/__init__.py", line 45, in <module>
    from rogerthat.consts import OFFICIALLY_SUPPORTED_LANGUAGES, DEBUG, FAST_QUEUE
  File "/Users/ruben/mobicage/git/rogerthat-backend/src/rogerthat/consts.py", line 22, in <module>
    MC_DASHBOARD = users.User(u'dashboard@rogerth.at')
  File "/Users/ruben/mobicage/git/rogerthat-backend/src/add_1_monkey_patches.py", line 68, in __init__
    original_constructor(self, *args, **kwargs)
  File "/Users/ruben/google-cloud-sdk/platform/google_appengine/google/appengine/api/users.py", line 102, in __init__
    assert _auth_domain
AssertionError
ERROR: Module: test_monkey_patch could not be imported (file: /Users/ruben/mobicage/git/rogerthat-backend/src-test/rogerthat_tests/test_monkey_patch.py).
```

File: src/rogerthat/bizz/messaging.py [hunk 3]
```diff
@@ -3434,19 +3433,31 @@ def validate_pay(widget_to):
...
+    base_method = filter(lambda m: m.provider_id == widget_to.base_method.provider_id
+                                   and m.currency == widget_to.base_method.currency, widget_to.methods)
+    if not base_method:
+        msg = 'No method found with provider_id %s and currency %s' % (widget_to.base_method.provider_id,
+                                                                       widget_to.base_method.currency)
+        raise InvalidValueException('base_method', msg)
+
+    currency_provider_map = {}
     for method in widget_to.methods:
+        logging.warn(method)
```
Comment (RubenMattan) [inline src/rogerthat/bizz/messaging.py]: debug logging

File: src/rogerthat/bizz/payment/__init__.py [hunk 4]
```diff
@@ -180,6 +173,20 @@ def update_payment_provider(provider_id, data):
...
+@returns()
+@arguments(provider_id=unicode)
+def delete_payment_provider(provider_id):
+    provider = get_payment_provider(provider_id)
+    if not provider:
+        raise PaymentProviderNotFoundException(provider_id)
+    to_delete = [provider.key]
+    to_delete.extend(PaymentUser.list_by_provider_id(provider_id).fetch(keys_only=True))
+    to_delete.extend(PaymentService.list_by_provider_id(provider_id, True).fetch(keys_only=True))
+    to_delete.extend(PaymentService.list_by_provider_id(provider_id, False).fetch(keys_only=True))
```
Comment (RubenMattan) [inline src/rogerthat/bizz/payment/__init__.py]: You can't delete the whole model but rather remove only the provider

File: src/rogerthat/bizz/payment/providers/payconiq/api.py [hunk 5]
```diff
@@ -228,9 +228,9 @@ def sync():
 @arguments(app_user=users.User, params=unicode)
 def create_transaction(app_user, params):
     payload = json.loads(params)
+    logging.info('Creating transaction with payload %s', payload)
```
Comment (RubenMattan) [inline src/rogerthat/bizz/payment/providers/payconiq/api.py]: This info is already logged by the expose(api) itself

File: src/rogerthat/dal/app.py [hunk 6]
```diff
@@ -102,6 +103,8 @@ def get_app_by_user(app_user):
 @returns(App)
 @arguments()
 def get_default_app():
+    if DEBUG:
+        return App.get_by_key_name('rogerthat')
```
Comment (RubenMattan) [inline src/rogerthat/dal/app.py]: debug?

File: src/rogerthat/models/payment.py [hunk 7]
```diff
@@ -101,12 +113,27 @@ def redirect_url(self, base_url):
...
         value = self.settings.get(setting)
         if not value:
             raise ValueError('PaymentProvider %s setting %s is not set' % (self.id, setting))
         return value
 
+    def get_amount(self, base_currency_amount, method):
+        # type: (float, PaymentMethodTO) -> long
+        if not base_currency_amount:
+            return method.amount
+        return long(round(base_currency_amount * self.get_currency_rate(method.currency) * pow(10, method.precision)))
```
Comment (RubenMattan) [inline src/rogerthat/models/payment.py]: it should be (amount / precision) * rate

File: src/rogerthat/models/properties/forms.py [hunk 8]
```diff
@@ -620,34 +626,49 @@ def deserialize_oauth(stream):
 
 
 def _serialize_payment_method(stream, c):
-    s_long(stream, 1)  # version
+    s_long(stream, 2)  # version
     s_unicode(stream, c.provider_id)
     s_unicode(stream, c.currency)
     s_long(stream, c.amount)
     s_long(stream, c.precision)
+    s_bool(stream, c.calculate_amount)
```
Comment (RubenMattan) [inline src/rogerthat/models/properties/forms.py]: target is not serialized

File: src/rogerthat/bizz/payment/__init__.py [hunk 9]
```diff
@@ -180,6 +172,40 @@ def update_payment_provider(provider_id, data):
...
+        return thing.provider_id != provider_id
+
+    for payment_user in PaymentUser.list_by_provider_id(provider_id):
+        payment_user.providers = filter(filter_by_provider, payment_user.providers)
+        payment_user.assets = filter(filter_by_provider, payment_user.assets)
+        if not payment_user.providers and not payment_user.assets:
+            to_delete.append(payment_user.key)
+        else:
+            to_put.append(payment_user)
+    services = PaymentService.list_by_provider_id(provider_id, True).fetch() + PaymentService.list_by_provider_id(
```
Comment (RubenMattan) [inline src/rogerthat/bizz/payment/__init__.py]: Refactor to 3 run jobs, this will fail with lots of users.

**Comment by lucasvanhalst** [comment]
let's wait until pay via chat is finished

**Pull request closed without merging**